[Inductor] Build Shape Padding in Inductor

jiawenliu64 opened the pull request:
Summary: Build shape padding for matmul/bmm/addmm in Inductor

Differential Revision: D41071282



cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

eellison commented on torch/_inductor/config.py:
```diff
@@ -67,6 +67,9 @@
 dynamo_import = inductor_import.replace("inductor", "dynamo")
 
+# Pad input tensors of matmul/bmm/addmm to leverage Tensor Cores in NVIDIA GPUs
+shape_padding = os.environ.get("TORCHINDUCTOR_SHAPE_PADDING", "0") == "1"
```
what's stopping us from turning this on by default ?

jiawenliu64 replied:
Hi @eellison, we tested shape padding optimization in a few internal and OSS models, and we will make it by default when the optimization is
- tested in most internal and OSS benchmarking models;
- supported other dtype (e.g., FP16 for inference) with different heuristics

jiawenliu64 commented:
@pytorchmergebot merge -f “macos test failed also on the trunk"

jiawenliu64 commented:
@pytorchbot rebase -s

jiawenliu64 commented:
@pytorchbot merge -f 'all failures are also on the trunk'

jiawenliu64 commented:
@pytorchbot merge -f 'all failures are also on the trunk'

jiawenliu64 commented:
@pytorchbot rebase -s

jgong5 commented on torch/_inductor/decomposition.py:
```diff
@@ -144,8 +167,138 @@ def addmm(input, mat1, mat2, *, beta=1, alpha=1):
+    return NotImplemented  # go directly to lowering
+
+
+def bench(f, name=None, iters=100, warmup=5, display=False):
```
Try to reuse the bench code in utils.py?
https://github.com/pytorch/pytorch/blob/master/torch/_inductor/utils.py#L106

jiawenliu64 replied:
In `utils.py`, we add a new bench which is more accurate and reliable

Chillee commented on torch/_inductor/decomposition.py:
```diff
@@ -144,8 +167,138 @@ def addmm(input, mat1, mat2, *, beta=1, alpha=1):
+            mat2 = pad_dim(mat2, k_padded_length, 1)
+        if n_padded_length != 0:
+            mat2 = pad_dim(mat2, n_padded_length, 2)
+            return torch.ops.aten.bmm(mat1, mat2)[:, :, :-n_padded_length].contiguous()
```
I added this contiguous call sinceI ran into needing it for correctness, but I think this might actually necessary in all cases?

cc: @ngimel 

It does sadly seem to impact perf a bit in my testing though.

Chillee reviewed (approved):
LGTM overall

Commit:
[Inductor] Build Shape Padding in Inductor (#88709)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/88709

Build shape padding for matmul/bmm/addmm in Inductor

Reviewed By: bertmaher, Chillee

Differential Revision: D41071282

fbshipit-source-id: e934f5c1523661b976800ea33a6c19e846147e4c

jiawenliu64 commented:
@pytorchbot merge

jiawenliu64 commented:
@pytorchbot merge -f "test failure unrelated"
